### DISCUSSION TITLE
Populating aux. variable from a function

### DISCUSSION THREAD

**Discussion by am-tc01** [idaholab/moose#18292] [Q&A General] [answered]
Hi,
I want to update an auxiliary variable (T) using a function that is dependent on a variable I am solving for (u). The function should point to an input data file that has two columns, values of u in the first column and T in the second. Furthermore, I want to get the values on nodal points and not a volume-averaged value on elemental level. Is there a straightforward way to do it?
Thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

You should look into AuxKernels to do this. 
If the function from u to T is simple, you can use a ParsedAux
If not, you should modify derive from the SelfAux for your class. You will have to create your own as there are none that do this tabulation you need afaik. 

If the AuxVariable is a nodal variable and so is u, that should not be a problem.

Guillaume

**Reply by am-tc01** [reply]
Hi Guillaume

Thanks for the reply. Yes, I've been looking into Auxkernels but couldn't find any kernel that I could just use off-the-shelf. FunctionAux could be one but then I can only have a function of x,y,z or t. Also, the function form u to T is not simple so I have to read in the data from an input file.
Thanks!

**Reply by GiudGiud** [reply]
FunctionAux could work with a ParsedFunction for your application, but if the function is not simple then that would not work because you have to be able to write down the expression for it in the input file. If you are ok with a really long input file, you could use python to write the table in a functionalized 'if u in this interval etc' form. 

I think you've identified a need. If you want to implement this capability, I could see it either as an AuxKernel, something like PiecewiseConstantAux OR more naturally as an extension of the Piecewise-... functions, which would let you specify a variable in addition to the axis eg x,y,z,t
